Fill Cucumber-Rust page with information from website (#324)
cucumber/docs: content/installation/rust.md

weight: 1200
---

> ## Cucmber-Rust

Reviewer: Build fails because: "First header must be a h1, but it was <h2".
Could you please change your headers to start with H1, and so on?

Author: Fixed

> Cucumber-Rust is an unofficial implementation of the Cucumber testing framework for Rust. Fully native, no external test runners or dependencies.

Reviewer: ```suggestion
Cucumber-Rust is an unofficial implementation of the Cucumber testing framework for Rust. The implementation is fully native. It doesn't use any external test runners or dependencies.
```

### Usage

Create a directory called `tests/` in your project root and create a test target of your choice. In this example we will name it `cucumber.rs`.

> Add this to your `Cargo.toml`:

Reviewer 2: Has this file been auto-generated somehow above that I've missed....

For example in ruby you can do `cucumber --init`

Author: In this case, you should do it by hands. I changed description a little to make it more clear.

```toml
[[test]]
name = "cucumber"
harness = false # Allows Cucumber to print output instead of libtest

[dev-dependencies]
cucumber_rust = "^0.5.1"
```

> If using stable Rust, you need to disable the `nightly` feature:

Reviewer: ```suggestion
If you are using stable Rust, you need to disable the nightly feature:
```

```
cucumber_rust = { version = "^0.5.1", default-features = false }
```

> Create a directory called `features/` and put a feature file in it named something like `example.feature`. It might look like:

Reviewer: ```suggestion
Create a directory called `features/` and place your feature file there. For example, a file named `example.feature`, which might look like this:
```

```gherkin
Feature: Example feature

  Scenario: An example scenario
    Given I am trying out Cucumber
    When I consider what I am doing

>     Then I am interested in ATDD

Reviewer: Is there a reason you are referring to ATDD instead of BDD?

Author: This is example from official repo

Reviewer: OK, thanks :)

    And we can implement rules with regex

```

> And here's an example of implementing those steps using our `tests/cucumber.rs` file:

Reviewer: ```suggestion
Here's an example of how to implement those steps using our `tests/cucumber.rs` file:
```

});

// A setup function to be called before everything else
fn setup() {
    
}

cucumber! {
    features: "./features", // Path to our feature files
    world: ::MyWorld, // The world needs to be the same for steps and the main cucumber call
    steps: &[
        example_steps::steps // the `steps!` macro creates a `steps` function in a module
    ],
    setup: setup, // Optional; called once before everything
    before: &[
        a_before_fn // Optional; called before each scenario
    ], 
    after: &[
        an_after_fn // Optional; called after each scenario
    ] 
}
```

The `cucumber!` creates the `main` function to be run.

The `steps!` macro generates a function named `steps` with all the declared steps in the module
it is defined in. Ordinarily you would create something like a `steps/` directory to hold your 
steps modules instead of inline like the given example.

The full gamut of Cucumber's Gherkin language is implemented by the 
[gherkin-rust](https://github.com/bbqsrc/gherkin-rust) project. Most features of the Gherkin 
language are parsed already and accessible via the relevant structs.

### License

This project is licensed under either of

>  * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)

Reviewer: Please make sure to link to an actual page/URL here, as the build is failing on these links not working.

Reviewer: And tbh I don't know if licence needs to be mentioned here (assuming it's available in the project).

Author: Fixed
